packages/requisitions/src/SupplierRequisition/DetailView/reducer.ts

```diff
@@ -1,35 +1,55 @@
-import { DefaultDocumentAction } from './../../../../common/src/hooks/useDocument/types';
-import { SupplierRequisitionLine } from './../../types';
-
+import { SupplierRequisitionLine, RequisitionActionType } from './../../types';
+import { Dispatch } from 'react';
 import {
   produce,
   DocumentActionSet,
   DocumentActionType,
   SortBy,
 } from '@openmsupply-client/common';
 import { placeholderSupplierRequisition } from '../../utils';
-import { SupplierRequisition, Requisition } from '../../types';
+import {
+  RequisitionAction,
+  SupplierRequisition,
+  Requisition,
+} from '../../types';
+import { Name } from '../../../../mock-server/src/data/types';
 
 export interface SupplierRequisitionStateShape {
   draft: SupplierRequisition;
   sortBy: SortBy<SupplierRequisitionLine>;
 }
 
+const RequisitionActionCreator = {
+  update: (key: string, value: string): RequisitionAction => {
+    return {
+      type: RequisitionActionType.Update,
+      payload: { key, value },
+    };
+  },
+  updateOtherParty: (value: Name): RequisitionAction => {
+    return {
+      type: RequisitionActionType.UpdateOtherParty,
+      payload: { value },
+    };
+  },
```

Comment:
Spent a bit of time trying to iterate on the invoices solution.

I spent a bit too much time trying to come up with some types which could allow for a single `update` function, where the `key` and `value` were typed correctly... 

I created the union on types with:

```
export type UpdateActions<ActionType, EntityType> = {
  [K in keyof EntityType]: {
    type: ActionType;
    payload: { key: K; value: EntityType[K] };
  };
}[keyof EntityType];
```

It gets close, but then in the reducer the key and values are still a union of all types, of course, 

i.e. you still can't do `state.draft[action.payload.key] = action.payload.value` of course 😢 

So I wasn't sure what to do - think just using a unique action per field might just be the way to go. The motivation to try make a nice generic fn was for plugins - if they're passed the draft they can just update it

Reply:
argh. yes, I spent a while trying as well - will have another go later, the goal of having a generic function for plugins is worth it. that would be awesome if possible

Reply:
Yeah. Does depend on what you allow plugins to do though. If they're only editing some sort of `plugin` object, then it might not be as useful.

Other option could be to write a function to generate `updateFieldName` methods and types which might be easier 🤔